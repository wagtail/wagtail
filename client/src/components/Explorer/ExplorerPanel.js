import PropTypes from 'prop-types';
import React from 'react';
import FocusTrap from 'focus-trap-react';

import { STRINGS, MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

import Button from '../Button/Button';
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
import Transition, { PUSH, POP } from '../Transition/Transition';
import ExplorerHeader from './ExplorerHeader';
import ExplorerItem from './ExplorerItem';
import PageCount from './PageCount';

/**
 * The main panel of the page explorer menu, with heading,
 * menu items, and special states.
 */
class ExplorerPanel extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      transition: PUSH,
      paused: false,
    };

    this.onItemClick = this.onItemClick.bind(this);
    this.onHeaderClick = this.onHeaderClick.bind(this);
    this.clickOutside = this.clickOutside.bind(this);
  }

  componentWillReceiveProps(newProps) {
    const { path } = this.props;
    const isPush = newProps.path.length > path.length;

    this.setState({
      transition: isPush ? PUSH : POP,
    });
  }

  componentDidMount() {
    document.querySelector('[data-explorer-menu-item]').classList.add('submenu-active');
    document.body.classList.add('explorer-open');
    document.addEventListener('mousedown', this.clickOutside);
    document.addEventListener('touchend', this.clickOutside);
  }

  componentWillUnmount() {
    document.querySelector('[data-explorer-menu-item]').classList.remove('submenu-active');
    document.body.classList.remove('explorer-open');
    document.removeEventListener('mousedown', this.clickOutside);
    document.removeEventListener('touchend', this.clickOutside);
  }

  clickOutside(e) {
    const { onClose } = this.props;
    const explorer = document.querySelector('[data-explorer-menu]');
    const toggle = document.querySelector('[data-explorer-menu-item]');

    const isInside = explorer.contains(e.target) || toggle.contains(e.target);
    if (!isInside) {
      onClose();
    }

    if (toggle.contains(e.target)) {
      this.setState({
        paused: true,
      });
    }
  }

  onItemClick(id, e) {
    const { pushPage } = this.props;

    e.preventDefault();
    e.stopPropagation();

    pushPage(id);
  }

  onHeaderClick(e) {
    const { path, popPage } = this.props;
    const hasBack = path.length > 1;

    if (hasBack) {
      e.preventDefault();
      e.stopPropagation();

      popPage();
    }
  }

  renderChildren() {
    const { page, nodes } = this.props;
    let children;

    if (!page.isFetching && !page.children.items) {
      children = (
        <div key="empty" className="c-explorer__placeholder">
          {STRINGS.NO_RESULTS}
        </div>
      );
    } else {
      children = (
        <div key="children">
          {page.children.items.map((id) => (
            <ExplorerItem
              key={id}
              item={nodes[id]}
              onClick={this.onItemClick.bind(null, id)}
            />
          ))}
        </div>
      );
    }

    return (
      <div className="c-explorer__drawer">
        {children}
        {page.isFetching ? (
          <div key="fetching" className="c-explorer__placeholder">
            <LoadingSpinner />
          </div>
        ) : null}
        {page.isError ? (
          <div key="error" className="c-explorer__placeholder">
            {STRINGS.SERVER_ERROR}
          </div>
        ) : null}
      </div>
    );
  }

  render() {
    const { page, onClose, path } = this.props;
    const { transition, paused } = this.state;

    return (
      <FocusTrap
        tag="nav"
        className="explorer"
        paused={paused || !page || page.isFetching}
        focusTrapOptions={{
          initialFocus: '.c-explorer__close',
          onDeactivate: onClose,
        }}
      >
        <Button className="c-explorer__close u-hidden" onClick={onClose}>
          {STRINGS.CLOSE_EXPLORER}
        </Button>
        <Transition name={transition} className="c-explorer">
          <div key={path.length} className="c-transition-group">
            <ExplorerHeader
              depth={path.length}
              page={page}
              onClick={this.onHeaderClick}
            />

            {this.renderChildren()}

            {page.isError || page.children.items && page.children.count > MAX_EXPLORER_PAGES ? (
              <PageCount page={page} />
            ) : null}
          </div>
        </Transition>
      </FocusTrap>
    );
  }
}

ExplorerPanel.propTypes = {
  nodes: PropTypes.object.isRequired,
  path: PropTypes.array.isRequired,
  page: PropTypes.shape({
    isFetching: PropTypes.bool,
    children: PropTypes.shape({
      count: PropTypes.number,
      items: PropTypes.array,
    }),
  }).isRequired,
  onClose: PropTypes.func.isRequired,
  popPage: PropTypes.func.isRequired,
  pushPage: PropTypes.func.isRequired,
};

export default ExplorerPanel;
