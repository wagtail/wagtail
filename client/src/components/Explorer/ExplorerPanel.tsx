import React from 'react';
import FocusTrap from 'focus-trap-react';

import { gettext } from '../../utils/gettext';
import { MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

import Button from '../Button/Button';
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
import Transition, { PUSH, POP } from '../Transition/Transition';
import ExplorerHeader from './ExplorerHeader';
import ExplorerItem from './ExplorerItem';
import PageCount from './PageCount';
import { State as NodeState, PageState } from './reducers/nodes';

interface ExplorerPanelProps {
  nodes: NodeState;
  depth: number;
  page: PageState;
  onClose(): void;
  gotoPage(id: number, transition: number): void;
}

interface ExplorerPanelState {
  transition: typeof PUSH | typeof POP;
  paused: boolean;
}

/**
 * The main panel of the page explorer menu, with heading,
 * menu items, and special states.
 */
class ExplorerPanel extends React.Component<
  ExplorerPanelProps,
  ExplorerPanelState
> {
  constructor(props) {
    super(props);

    this.state = {
      transition: PUSH,
      paused: false,
    };

    this.onItemClick = this.onItemClick.bind(this);
    this.onHeaderClick = this.onHeaderClick.bind(this);
    this.clickOutside = this.clickOutside.bind(this);
  }

  componentWillReceiveProps(newProps) {
    const { depth } = this.props;
    const isPush = newProps.depth > depth;

    this.setState({
      transition: isPush ? PUSH : POP,
    });
  }

  componentDidMount() {
    document
      .querySelector('[data-explorer-menu-item]')
      ?.classList.add('submenu-active');
    document.body.classList.add('explorer-open');
    document.addEventListener('mousedown', this.clickOutside);
    document.addEventListener('touchend', this.clickOutside);
  }

  componentWillUnmount() {
    document
      .querySelector('[data-explorer-menu-item]')
      ?.classList.remove('submenu-active');
    document.body.classList.remove('explorer-open');
    document.removeEventListener('mousedown', this.clickOutside);
    document.removeEventListener('touchend', this.clickOutside);
  }

  clickOutside(e) {
    const { onClose } = this.props;
    const explorer = document.querySelector('[data-explorer-menu]');
    const toggle = document.querySelector('[data-explorer-menu-item]');

    if (!explorer || !toggle) {
      return;
    }

    const isInside = explorer.contains(e.target) || toggle.contains(e.target);
    if (!isInside) {
      onClose();
    }

    if (toggle.contains(e.target)) {
      this.setState({
        paused: true,
      });
    }
  }

  onItemClick(id, e) {
    const { gotoPage } = this.props;

    e.preventDefault();
    e.stopPropagation();

    gotoPage(id, 1);
  }

  onHeaderClick(e) {
    const { page, depth, gotoPage } = this.props;
    const parent = page.meta.parent?.id;

    // Note: Checking depth as well in case the user started deep in the tree
    if (depth > 0 && parent) {
      e.preventDefault();
      e.stopPropagation();

      gotoPage(parent, -1);
    }
  }

  renderChildren() {
    const { page, nodes } = this.props;
    let children;

    if (!page.isFetchingChildren && !page.children.items) {
      children = (
        <div key="empty" className="c-explorer__placeholder">
          {gettext('No results')}
        </div>
      );
    } else {
      children = (
        <div key="children">
          {page.children.items.map((id) => (
            <ExplorerItem
              key={id}
              item={nodes[id]}
              onClick={this.onItemClick.bind(null, id)}
            />
          ))}
        </div>
      );
    }

    return (
      <div className="c-explorer__drawer">
        {children}
        {page.isFetchingChildren || page.isFetchingTranslations ? (
          <div key="fetching" className="c-explorer__placeholder">
            <LoadingSpinner />
          </div>
        ) : null}
        {page.isError ? (
          <div key="error" className="c-explorer__placeholder">
            {gettext('Server Error')}
          </div>
        ) : null}
      </div>
    );
  }

  render() {
    const { page, onClose, depth, gotoPage } = this.props;
    const { transition, paused } = this.state;

    return (
      <FocusTrap
        paused={
          paused ||
          !page ||
          page.isFetchingChildren ||
          page.isFetchingTranslations
        }
        focusTrapOptions={{
          initialFocus: '.c-explorer__header__title',
          onDeactivate: onClose,
        }}
      >
        <div role="dialog" className="explorer">
          <Button className="c-explorer__close">
            {gettext('Close explorer')}
          </Button>
          <Transition
            name={transition}
            className="c-explorer"
            component="nav"
            label={gettext('Page explorer')}
          >
            <div key={depth} className="c-transition-group">
              <ExplorerHeader
                depth={depth}
                page={page}
                onClick={this.onHeaderClick}
                gotoPage={gotoPage}
              />

              {this.renderChildren()}

              {page.isError ||
              (page.children.items &&
                page.children.count > MAX_EXPLORER_PAGES) ? (
                <PageCount page={page} />
              ) : null}
            </div>
          </Transition>
        </div>
      </FocusTrap>
    );
  }
}

export default ExplorerPanel;
