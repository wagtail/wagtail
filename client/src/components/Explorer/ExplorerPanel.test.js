import React from 'react';
import { shallow } from 'enzyme';
import ExplorerPanel from './ExplorerPanel';

const mockProps = {
  page: {
    children: {
      items: [],
    },
    meta: {
      parent: null,
    },
  },
  depth: 1,
  onClose: jest.fn(),
  gotoPage: jest.fn(),
  nodes: {},
};

describe('ExplorerPanel', () => {
  describe('general rendering', () => {
    beforeEach(() => {
      document.body.innerHTML = '<div data-explorer-menu-item></div>';
    });

    it('exists', () => {
      expect(ExplorerPanel).toBeDefined();
    });

    it('renders', () => {
      expect(shallow(<ExplorerPanel {...mockProps} />)).toMatchSnapshot();
    });

    it('#isFetching', () => {
      expect(
        shallow(
          <ExplorerPanel
            {...mockProps}
            page={Object.assign({ isFetching: true }, mockProps.page)}
          />,
        ),
      ).toMatchSnapshot();
    });

    it('#isError', () => {
      expect(
        shallow(
          <ExplorerPanel
            {...mockProps}
            page={Object.assign({ isError: true }, mockProps.page)}
          />,
        ),
      ).toMatchSnapshot();
    });

    it('no children', () => {
      expect(
        shallow(<ExplorerPanel {...mockProps} page={{ children: {} }} />),
      ).toMatchSnapshot();
    });

    it('#items', () => {
      expect(
        shallow(
          <ExplorerPanel
            {...mockProps}
            page={{ children: { items: [1, 2] } }}
            nodes={{
              1: {
                id: 1,
                admin_display_title: 'Test',
                meta: { status: {}, type: 'test' },
              },
              2: {
                id: 2,
                admin_display_title: 'Foo',
                meta: { status: {}, type: 'foo' },
              },
            }}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('onHeaderClick', () => {
    beforeEach(() => {
      mockProps.gotoPage.mockReset();
    });

    it('calls gotoPage', () => {
      shallow(
        <ExplorerPanel
          {...mockProps}
          depth={2}
          page={{ children: { items: [] }, meta: { parent: { id: 1 } } }}
        />,
      )
        .find('ExplorerHeader')
        .prop('onClick')({
        preventDefault() {},
        stopPropagation() {},
      });

      expect(mockProps.gotoPage).toHaveBeenCalled();
    });

    it('does not call gotoPage for first page', () => {
      shallow(
        <ExplorerPanel
          {...mockProps}
          depth={0}
          page={{ children: { items: [] }, meta: { parent: { id: 1 } } }}
        />,
      )
        .find('ExplorerHeader')
        .prop('onClick')({
        preventDefault() {},
        stopPropagation() {},
      });

      expect(mockProps.gotoPage).not.toHaveBeenCalled();
    });
  });

  describe('onItemClick', () => {
    beforeEach(() => {
      mockProps.gotoPage.mockReset();
    });

    it('calls gotoPage', () => {
      shallow(
        <ExplorerPanel
          {...mockProps}
          path={[1]}
          page={{ children: { items: [1] } }}
          nodes={{
            1: {
              id: 1,
              admin_display_title: 'Test',
              meta: { status: {}, type: 'test' },
            },
          }}
        />,
      )
        .find('ExplorerItem')
        .prop('onClick')({
        preventDefault() {},
        stopPropagation() {},
      });

      expect(mockProps.gotoPage).toHaveBeenCalled();
    });
  });

  describe('hooks', () => {
    it('componentWillReceiveProps push', () => {
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      expect(wrapper.setProps({ depth: 2 }).state('transition')).toBe('push');
    });

    it('componentWillReceiveProps pop', () => {
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      expect(wrapper.setProps({ depth: 0 }).state('transition')).toBe('pop');
    });

    it('componentDidMount', () => {
      document.body.innerHTML = '<div data-explorer-menu-item></div>';
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      wrapper.instance().componentDidMount();
      expect(
        document
          .querySelector('[data-explorer-menu-item]')
          .classList.contains('submenu-active'),
      ).toBe(true);
      expect(document.body.classList.contains('explorer-open')).toBe(true);
    });

    it('componentWillUnmount', () => {
      document.body.innerHTML =
        '<div class="submenu-active" data-explorer-menu-item></div>';
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      wrapper.instance().componentWillUnmount();
      expect(
        document
          .querySelector('[data-explorer-menu-item]')
          .classList.contains('submenu-active'),
      ).toBe(false);
      expect(document.body.classList.contains('explorer-open')).toBe(false);
    });
  });

  describe('clickOutside', () => {
    afterEach(() => {
      mockProps.onClose.mockReset();
    });

    it('triggers onClose when click is outside', () => {
      document.body.innerHTML =
        '<div data-explorer-menu-item></div><div data-explorer-menu></div><div id="t"></div>';
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      wrapper.instance().clickOutside({
        target: document.querySelector('#t'),
      });
      expect(mockProps.onClose).toHaveBeenCalled();
    });

    it('does not trigger onClose when click is inside', () => {
      document.body.innerHTML =
        '<div data-explorer-menu-item></div><div data-explorer-menu><div id="t"></div></div>';
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      wrapper.instance().clickOutside({
        target: document.querySelector('#t'),
      });
      expect(mockProps.onClose).not.toHaveBeenCalled();
    });

    it('pauses focus trap inside toggle', () => {
      document.body.innerHTML =
        '<div data-explorer-menu-item><div id="t"></div></div><div data-explorer-menu></div>';
      const wrapper = shallow(<ExplorerPanel {...mockProps} />);
      wrapper.instance().clickOutside({
        target: document.querySelector('#t'),
      });
      expect(wrapper.state('paused')).toEqual(true);
    });
  });
});
