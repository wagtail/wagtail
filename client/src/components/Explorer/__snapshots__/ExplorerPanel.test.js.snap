// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerPanel general rendering #isError 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  className="explorer"
  focusTrapOptions={
    Object {
      "initialFocus": ".c-explorer__header",
      "onDeactivate": [MockFunction],
    }
  }
  paused={false}
  role="dialog"
  tag="div"
>
  <Button
    accessibleLabel={null}
    className="c-explorer__close"
    dialogTrigger={false}
    href="#"
    icon=""
    isLoading={false}
    onClick={[MockFunction]}
    preventDefault={true}
    target={null}
  >
    Close explorer
  </Button>
  <Transition
    className="c-explorer"
    component="nav"
    duration={210}
    label="Page explorer"
    name="push"
  >
    <div
      className="c-transition-group"
      key="0"
    >
      <ExplorerHeader
        depth={0}
        onClick={[Function]}
        page={
          Object {
            "children": Object {
              "items": Array [],
            },
            "isError": true,
          }
        }
      />
      <div
        className="c-explorer__drawer"
      >
        <div
          key="children"
        />
        <div
          className="c-explorer__placeholder"
          key="error"
        >
          Server Error
        </div>
      </div>
      <PageCount
        page={
          Object {
            "children": Object {
              "items": Array [],
            },
            "isError": true,
          }
        }
      />
    </div>
  </Transition>
</FocusTrap>
`;

exports[`ExplorerPanel general rendering #isFetching 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  className="explorer"
  focusTrapOptions={
    Object {
      "initialFocus": ".c-explorer__header",
      "onDeactivate": [MockFunction],
    }
  }
  paused={true}
  role="dialog"
  tag="div"
>
  <Button
    accessibleLabel={null}
    className="c-explorer__close"
    dialogTrigger={false}
    href="#"
    icon=""
    isLoading={false}
    onClick={[MockFunction]}
    preventDefault={true}
    target={null}
  >
    Close explorer
  </Button>
  <Transition
    className="c-explorer"
    component="nav"
    duration={210}
    label="Page explorer"
    name="push"
  >
    <div
      className="c-transition-group"
      key="0"
    >
      <ExplorerHeader
        depth={0}
        onClick={[Function]}
        page={
          Object {
            "children": Object {
              "items": Array [],
            },
            "isFetching": true,
          }
        }
      />
      <div
        className="c-explorer__drawer"
      >
        <div
          key="children"
        />
        <div
          className="c-explorer__placeholder"
          key="fetching"
        >
          <LoadingSpinner />
        </div>
      </div>
    </div>
  </Transition>
</FocusTrap>
`;

exports[`ExplorerPanel general rendering #items 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  className="explorer"
  focusTrapOptions={
    Object {
      "initialFocus": ".c-explorer__header",
      "onDeactivate": [MockFunction],
    }
  }
  paused={false}
  role="dialog"
  tag="div"
>
  <Button
    accessibleLabel={null}
    className="c-explorer__close"
    dialogTrigger={false}
    href="#"
    icon=""
    isLoading={false}
    onClick={[MockFunction]}
    preventDefault={true}
    target={null}
  >
    Close explorer
  </Button>
  <Transition
    className="c-explorer"
    component="nav"
    duration={210}
    label="Page explorer"
    name="push"
  >
    <div
      className="c-transition-group"
      key="0"
    >
      <ExplorerHeader
        depth={0}
        onClick={[Function]}
        page={
          Object {
            "children": Object {
              "items": Array [
                1,
                2,
              ],
            },
          }
        }
      />
      <div
        className="c-explorer__drawer"
      >
        <div
          key="children"
        >
          <ExplorerItem
            item={
              Object {
                "admin_display_title": "Test",
                "id": 1,
                "meta": Object {
                  "status": Object {},
                  "type": "test",
                },
              }
            }
            key="1"
            onClick={[Function]}
          />
          <ExplorerItem
            item={
              Object {
                "admin_display_title": "Foo",
                "id": 2,
                "meta": Object {
                  "status": Object {},
                  "type": "foo",
                },
              }
            }
            key="2"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </Transition>
</FocusTrap>
`;

exports[`ExplorerPanel general rendering no children 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  className="explorer"
  focusTrapOptions={
    Object {
      "initialFocus": ".c-explorer__header",
      "onDeactivate": [MockFunction],
    }
  }
  paused={false}
  role="dialog"
  tag="div"
>
  <Button
    accessibleLabel={null}
    className="c-explorer__close"
    dialogTrigger={false}
    href="#"
    icon=""
    isLoading={false}
    onClick={[MockFunction]}
    preventDefault={true}
    target={null}
  >
    Close explorer
  </Button>
  <Transition
    className="c-explorer"
    component="nav"
    duration={210}
    label="Page explorer"
    name="push"
  >
    <div
      className="c-transition-group"
      key="0"
    >
      <ExplorerHeader
        depth={0}
        onClick={[Function]}
        page={
          Object {
            "children": Object {},
          }
        }
      />
      <div
        className="c-explorer__drawer"
      >
        <div
          className="c-explorer__placeholder"
          key="empty"
        >
          No results
        </div>
      </div>
    </div>
  </Transition>
</FocusTrap>
`;

exports[`ExplorerPanel general rendering renders 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  className="explorer"
  focusTrapOptions={
    Object {
      "initialFocus": ".c-explorer__header",
      "onDeactivate": [MockFunction],
    }
  }
  paused={false}
  role="dialog"
  tag="div"
>
  <Button
    accessibleLabel={null}
    className="c-explorer__close"
    dialogTrigger={false}
    href="#"
    icon=""
    isLoading={false}
    onClick={[MockFunction]}
    preventDefault={true}
    target={null}
  >
    Close explorer
  </Button>
  <Transition
    className="c-explorer"
    component="nav"
    duration={210}
    label="Page explorer"
    name="push"
  >
    <div
      className="c-transition-group"
      key="0"
    >
      <ExplorerHeader
        depth={0}
        onClick={[Function]}
        page={
          Object {
            "children": Object {
              "items": Array [],
            },
          }
        }
      />
      <div
        className="c-explorer__drawer"
      >
        <div
          key="children"
        />
      </div>
    </div>
  </Transition>
</FocusTrap>
`;
